x/archivedb/db.go

```diff
@@ -0,0 +1,150 @@
+package archivedb
+
+import (
+	"bytes"
+	"context"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+)
+
+// ArchiveDb
+//
+// Creates a thin database layer on top of database.Database. ArchiveDb is an
+// append only database which stores all state changes happening at every block
+// height. Each record is stored in such way to perform both fast inserts and selects.
+//
+// Currently its API is quite simple, it has two main functions, one to create a
+// Batch write with a given block height, inside this batch entries can be added
+// with a given value or they can be deleted. It also provides a Get function
+// that takes a given key and a height.
+//
+//	The way it works is as follows:
+//		- Height: 10
+//			Set(foo, "foo's value is bar")
+//			Set(bar, "bar's value is bar")
+//		- Height: 100
+//			Set(foo, "updatedfoo's value is bar")
+//		- Height: 1000
+//			Set(bar, "updated bar's value is bar")
+//			Delete(foo)
+//
+// When requesting `Get(foo, 9)` it will return an errNotFound error because foo
+// was not defined at block height 9, it was defined later. When calling
+// `Get(foo, 99)` it will return a tuple `("foo's value is bar", 10)` returning
+// the value of `foo` at height 99 (which was set at height 10). If requesting
+// `Get(foo, 2000)` it will return an error because `foo` was deleted at height
+// 1000.
+type archiveDB struct {
+	// Must be held when reading/writing fields.
+	lock sync.RWMutex
+
+	ctx context.Context
+
+	rawDB database.Database
+}
+
+type batchWithHeight struct {
+	db     *archiveDB
+	height uint64
+	batch  database.Batch
+}
+
+func newDatabase(
+	ctx context.Context,
+	db database.Database,
+) (*archiveDB, error) {
+	return &archiveDB{
+		ctx:   ctx,
+		rawDB: db,
+	}, nil
+}
+
+// Fetches the value of a given prefix at a given height.
+//
+// If the value does not exists or it was actually removed an error is returned.
+// Otherwise a value does exists it will be returned, alongside with the height
+// at which it was updated prior the requested height.
+func (db *archiveDB) Get(key []byte, height uint64) ([]byte, uint64, error) {
```

Comment:
Very curious what the benchmarked performance on this is.

Wondering if it makes sense to add a cache to avoid iteration for commonly used key + heights?

Reply:
@patrick-ogrady The iterator will consume just a single entry at most. I don't think it would be overkill to be honest; I will create some benchmark scripts with some real load and many million entries to measure performance.

Reply:
Could we also add the golang style benchmarks?

Reply:
I'd be in favor of making any caching a follow-on PR